text/0300-programmatic-toolkit.md

```diff
@@ -0,0 +1,776 @@
+> If we end up implementing a lot of transactional retries, we can revisit this decision and add recoverable errors as a generic reusable way.
+
+While most errors are terminal, it's possible to recover from some.
+For example if an authentication token is expired, an integrator may prompt the user to re-authenticate via a browser login, before continuing with the action.
+With recoverable errors, the integrator can return to normal program flow after an issue has been addressed.
+
+In the [Programmatic Toolkit], recoverable errors are implemented as a special type of request.
+By default, errors are not attempted to recover.
+
+```ts
+interface RecoverableError<T, U> extends IoRequest<T, U> {
+  level: 'error';
+  defaultResponse: {
+    retry: boolean; // indicates if the block should retried
+    // ... other response values
+  };
+  isRecoverable: true;
+  attempt: number; // the number of the current attempt, starting with 1
+}
+```
+
+When `IoHost.requestResponse()` is called with a recoverable error, the integrator may choose to return `retry: true` to indicate the block should be retried.
+A single retry is allowed, if the block fails again with the same error it will be raised as an exception.
+Retries are not guaranteed and integrators must handle the case where a requested retry is not executed.
+Recoverable errors are rare as they require special programming and not always possible.
+
+### What are the drawbacks of this solution?
+
+#### Increased API surface
+
+The main risk is the larger surface area.
+Previously we have only committed to the CLI interface, which is fairly limited by design.
+The new API surface covers additional option interfaces, emitted messages and newly introduced primitives such as a `CloudExecutable` or `StackDeployment`.
+
+Our primary mitigation is to constrain our the new APIs to real world end-to-end use cases (see tenet T2).s
+
+#### Separating actions and interaction makes it harder to add features to the CLI
+
+The solution includes a paradigm shift to consider programmatic access first and the CLI second.
+This shift is codified by the introduction of a new package that contains all actions and the CLI calling them.
+With this, we now have more work to do: Any feature needs to first consider its programmatic usage and then derive the CLI usage from it.
+
+We mitigate this by providing tooling that helps contributors to define different inputs and defaults declaratively and generate (parts of) the code for it.
+Additional test tooling will help writing the same test cases onces for different input modes.
+
+#### A lot of refactoring
+
+The proposal requires to move a lot of code around, mostly into new packages.
+It also implies refactoring of existing interfaces to match the new API design.
+Both of these things are prone to human errors.
+
+Mitigation for this is test coverage and separating out PRs that are purely moving code, from actual changes.
```

Comment:
We probably want to coordinate rolling this out only after we address the coverage gaps we detected.